mmv1/third_party/terraform/services/compute/data_source_google_compute_reservation.go

```diff
@@ -0,0 +1,114 @@
+package compute
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/terraform-provider-google/google/tpgresource"
+	transport_tpg "github.com/hashicorp/terraform-provider-google/google/transport"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+var (
+	computeReservationIdTemplate = "projects/%s/zones/%s/reservations/%s"
+	computeReservationLinkRegex  = regexp.MustCompile("projects/(.+)/zones/(.+)/reservations/(.+)$")
+)
+
+func DataSourceGoogleComputeReservation() *schema.Resource {
+	// Generate datasource schema from resource
+	dsSchema := tpgresource.DatasourceSchemaFromResourceSchema(ResourceComputeReservation().Schema)
+
+	// Set 'Required' schema elements
+	tpgresource.AddRequiredFieldsToSchema(dsSchema, "name")
+	tpgresource.AddRequiredFieldsToSchema(dsSchema, "zone")
+
+	tpgresource.AddOptionalFieldsToSchema(dsSchema, "project")
+
+	return &schema.Resource{
+		Read:   dataSourceGoogleComputeReservationRead,
+		Schema: dsSchema,
+	}
+}
+
+func dataSourceGoogleComputeReservationRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*transport_tpg.Config)
+	err := resourceComputeReservationRead(d, meta)
+	if err != nil {
+		return err
+	}
+
+	project, err := tpgresource.GetProject(d, config)
+	if err != nil {
+		return err
+	}
+	zone, err := tpgresource.GetZone(d, config)
+	if err != nil {
+		return err
+	}
+	name := d.Get("name").(string)
+
+	d.SetId(fmt.Sprintf("projects/%s/zones/%s/reservations/%s", project, zone, name))
+	return nil
+}
+
+type ComputeReservationId struct {
+	Project string
+	Zone    string
+	Name    string
+}
+
+func (s ComputeReservationId) CanonicalId() string {
+	return fmt.Sprintf(computeReservationIdTemplate, s.Project, s.Zone, s.Name)
+}
+
+func ParseComputeReservationId(id string, config *transport_tpg.Config) (*ComputeReservationId, error) {
```

Comment:
In general, data sources shouldn't need a function like this. I see that Compute Address has something similar, which may be where this is coming from? But it's not a common pattern - especially if it's test-only.

Would it be possible to remove this function and do the tests some other way?

Reply:
Looked into what affects it would have if we removed `ParseComputeReservationId`, turns out it's not necessary at all and was really only added because of a similar function being used in compute address like you mentioned.

I've removed it to keep the test simple.